### PR TITLE
Miniostoring

### PR DESCRIPTION
**Pull request by mhymhygi**
A provider can store BLOBs in [Minio Object Storing](https://min.io/)

### CODE REVIEW

**Comment by liangshiw** [comment]
Hi @mhymhygi,

Your PR contains some unwanted changes.

image

**Comment by mhymhygi** [comment]
Sorry,my mistake.I will remove those.




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "liangshiwei"<notifications@github.com&gt;; 
发送时间: 2020年7月9日(星期四) 上午10:17
收件人: "abpframework/abp"<abp@noreply.github.com&gt;; 
抄送: "Biserrata"<274871620@qq.com&gt;; "Mention"<mention@noreply.github.com&gt;; 
主题: Re: [abpframework/abp] Miniostoring (#4688)





 
Hi @mhymhygi,
 
Your PR contains some unwanted changes.

File: framework/src/Volo.Abp.BlobStoring.Minio/Volo/Abp/BlobStoring/Minio/MinioBlobProvider.cs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        {
+            // Make sure Blob Container exists.
+            if (await ContainerExistsAsync(args))
+            {
+                try
+                {
+                    await GetMinioClient(args).StatObjectAsync(GetContainerName(args), blobName);
+                    return true;
+                }
+                catch (MinioException ex)
```
Comment (liangshiw) [inline framework/src/Volo.Abp.BlobStoring.Minio/Volo/Abp/BlobStoring/Minio/MinioBlobProvider.cs:134]: Why hide exception?
Developers never get exceptions,I think this is not a good practice. 
Just throw, that's fine.

**Comment by liangshiw** [comment]
Hi @mhymhygi 

I just simply view your code, I think you should make the following changes:

1. Keep code style consistent with ABP, like Indentation, distance between method and property etc ... Please refer Azure blob provider.
2. Remove unwanted try, catch code block.
3. I see there are some unused variables and I think they can be remove.

Thanks.

**Comment by mhymhygi** [comment]
Thank you for reviewing my code and your valuable comments.@liangshiw

**Comment by mhymhygi** [comment]
Hi,@liangshiw .
I had made some changes.
But there is the only way to check the object wether exists or not by catch the exception.
I do not have good idea.

File: framework/src/Volo.Abp.BlobStoring.Minio/Volo/Abp/BlobStoring/Minio/MinioBlobProvider.cs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            {
+                try
+                {
+                    // Check whether the object exists using statObject().
+                    // If the object is not found, statObject() throws an exception,
+                    // else it means that the object exists.
+                    // Execution is successful.
+                    await client.StatObjectAsync(containerName, blobName);
+                }
+                catch (Exception e)
```
Comment (liangshiw) [inline framework/src/Volo.Abp.BlobStoring.Minio/Volo/Abp/BlobStoring/Minio/MinioBlobProvider.cs:124]:
```csharp
try
{
    await client.StatObjectAsync(containerName, blobName);
}
catch (Exception e)
{
    if (e is ObjectNotFoundException)
    {
        return false;
    }

    throw;
}
```

File: framework/test/Volo.Abp.BlobStoring.Minio.Tests/Volo/Abp/BlobStoring/Minio/AbpBlobStoringMinioTestModule.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                        minio.SecretKey = _secretKey;
+                        minio.WithSSL = false;
+                        minio.BucketName = _randomContainerName;
+                        minio.CreateBucketIfNotExists = true;
+                    });
+                });
+            });
+        }
+
+        public async override void OnApplicationShutdown(ApplicationShutdownContext context)
```
Comment (liangshiw) [inline framework/test/Volo.Abp.BlobStoring.Minio.Tests/Volo/Abp/BlobStoring/Minio/AbpBlobStoringMinioTestModule.cs:62]: Should be  `public override async void OnApplicationShutdown(ApplicationShutdownContext context)`

File: framework/test/Volo.Abp.BlobStoring.Minio.Tests/Volo/Abp/BlobStoring/Minio/AbpBlobStoringMinioTestModule.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        {
+            var minioClient = new MinioClient(_endPoint, _accessKey, _secretKey);
+            if (await minioClient.BucketExistsAsync(_randomContainerName))
+            {
+                var observables =minioClient.ListObjectsAsync(_randomContainerName,null,true);
```
Comment (liangshiw) [inline framework/test/Volo.Abp.BlobStoring.Minio.Tests/Volo/Abp/BlobStoring/Minio/AbpBlobStoringMinioTestModule.cs:67]: Running unit tests will get exceptions.

image

Should change to:

```csharp
foreach (var item in objects)
{
    await minioClient.RemoveObjectAsync(_randomContainerName, item.Key);
}

await minioClient.RemoveBucketAsync(_randomContainerName);
```

File: .gitignore [hunk 5]
```diff
@@ -304,3 +304,11 @@ modules/virtual-file-explorer/app/Volo.Abp.VirtualFileExplorer.DemoApp/Logs/
 /templates/app/angular/package-lock.json
 /modules/client-simulation/demo/Volo.ClientSimulation.Demo/package-lock.json
 
+/framework/src/Volo.Abp.AspNetCore/Properties/launchSettings.json
```
Comment (liangshiw) [inline .gitignore:307]: Why add these?

**Comment by liangshiw** [comment]
In addition, we should normalize the container name, like https://github.com/abpframework/abp/blob/dev/framework/src/Volo.Abp.BlobStoring.Azure/Volo/Abp/BlobStoring/Azure/AzureBlobNamingNormalizer.cs.
About rules for bucket naming,please see : https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html

**Comment by liangshiw** [comment]
Thanks a lot @mhymhygi  👍
Can you write documentation for minio integration?

**Comment by mhymhygi** [comment]
Yes,I can do it.@liangshiw

**Pull request merged**